File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/entries/EntryPage.java
package in.org.quicko.silo.client.entries;

import in.org.quicko.silo.client.pagination.Page;
import in.org.quicko.silo.client.pagination.PageWindow;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;

/**
 * One page of entries, navigated by the window the server echoed rather than the
 * one requested.
 */
public final class EntryPage<F> extends Page<Entry<F>> {
  private final Function<PageWindow, EntryPage<F>> loader;

  public EntryPage(
      List<Entry<F>> rows, int total, PageWindow window, Function<PageWindow, EntryPage<F>> loader) {
    super(rows, total, window);
    this.loader = loader;
  }

  public List<Entry<F>> entries() {
    return rows();
  }

  public Optional<EntryPage<F>> next() {
    return windowForNext().map(loader);
  }

  public Optional<EntryPage<F>> previous() {
    return window().previous().map(loader);
  }
}

File: packages/silo-client-dart/lib/src/entries/entry_page.dart
import '../pagination/page.dart';
import '../pagination/page_window.dart';
import 'entry.dart';

/// One page of entries, navigated by the window the server answered.
final class EntryPage<F> extends Page<Entry<F>> {
  EntryPage(this.entries, int total, PageWindow window, this._loader) : super(entries, total, window);

  final List<Entry<F>> entries;
  final Future<EntryPage<F>> Function(PageWindow window) _loader;

  Future<EntryPage<F>?> next() async {
    final window = windowForNext();
    return window == null ? null : _loader(window);
  }

  Future<EntryPage<F>?> previous() async {
    final window = windowForPrevious();
    return window == null ? null : _loader(window);
  }
}
